src/solver/cat/forward_cat.cpp

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/cat/solvers.hpp>
+
+#include <miopen/cat/cat_invoke_params.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/cat.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 256
+#define MAX_TENSOR_X_COUNT 8
+#define MIN_OUTPUT_TENSOR_SIZE 1000000
+
+namespace miopen {
+
+namespace solver {
+
+namespace cat {
+
+bool IsUnderXCountLimit(const miopen::cat::ProblemDescription& problem)
+{
+    if(problem.GetXCount() > MAX_TENSOR_X_COUNT)
+    {
+        MIOPEN_THROW(miopenStatusBadParm, "CatForward: Exceeded the number of tensors.");
+    }
+    return true;
+}
+
+bool IsImprovementOverROCm(const miopen::cat::ProblemDescription& problem)
+{
+    if(problem.GetYDesc().GetElementSize() < MIN_OUTPUT_TENSOR_SIZE)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool CatForward::IsApplicable(const ExecutionContext& context,
+                              const miopen::cat::ProblemDescription& problem) const
+{
+    if(!IsUnderXCountLimit(problem))
+        return false;
+    if(!problem.IsSameType())
```

Comment:
IsSameType and some other functions  never returns false (it throws an exception instead). Is it intended?

Reply:
@Vsevolod1983 Good catch! #2639 opened.

Reply:
fix it, thank you.

Reply:
[resolved]